### PR TITLE
Added key transform annotations with tests

### PR DESCRIPTION
**Pull request by lbialy**
Adds annotations that allow to transform all keys to a predefined scheme, ie snake_case or PascalCase or even iNDIANAjONEScASE. Naming, deduplication of code between scala 2 and scala 3 and other stuff is open for discussion, I'd be glad to clear stuff up.

Scala 3 code is not tested, it's not even compiled when `sbt test` is called (or `sbt compile` for that matter). I've also found one thing that looked very weird - stuff like params or names were computed inside of unsafeEncode method in Scala 3 JsonEncoder, which means that those operations that are executed once on creation of JsonEn/Decoder for all other cases would be repeated for all calls to unsafeEncode. If that's how it's meant to be for some obscure reason I can revert that. I have no way of knowing whether that works because I don't know how to even force sbt to compile Scala 3 macros sources, sorry.

Cheers!

### CODE REVIEW

**Comment by lbialy** [comment]
Stuff should compile and test just fine now, documentation was updated to mention new feature. I think I have made JsonEncoder for Scala 3 a bit more performant by moving all the non-necessary work out of scope of `unsafeEncode` method (just to match how it was done in other parts of the codebase). Please shout if anything is borked in the diff.

File: zio-json/shared/src/main/scala-3/zio/json/macros.scala [hunk 1]
```diff
@@ -42,6 +42,93 @@ final case class jsonDiscriminator(name: String) extends Annotation
...
+/**
+ * If used on a case class, determines the strategy of member names
+ * transformation during serialization and deserialization. Four common
+ * strategies are provided above and a custom one to support specific use cases.
+ */
+final case class jsonMemberNames(format: JsonMemberFormat) extends Annotation
+private[json] object jsonMemberNames {
+
+  /**
+   * ~~Stolen~~ Borrowed from jsoniter-scala by Andriy Plokhotnyuk
```
Comment (fsvehla) [inline zio-json/shared/src/main/scala-3/zio/json/macros.scala:72]: Thanks, @plokhotnyuk

**Comment by fsvehla** [comment]
This looks good to me, @lbialy, thanks for the work!

I know that you’ve looked into a possible issue with inline defs, are you still ready to have this merged?

**Comment by runtologist** [comment]
@lbialy This look great. I'd love to use it. Are there plans for getting this merged?

**Comment by lbialy** [comment]
@fsvehla merged with develop, one word of warning - it seems that `testzio.json.JavaTimeSpec` is failing both on develop and now on this PR

**Comment by fsvehla** [comment]
@lbialy I can’t reproduce the failing `JavaTimeSpec`. Mind pasting the stacktrace?

File: zio-json/shared/src/main/scala-3/zio/json/macros.scala (hunk 1 above)
Comment (plokhotnyuk) [reply zio-json/shared/src/main/scala-3/zio/json/macros.scala:72]: I’m happy to help!

BTW, please give your GitHub stars to [jsoniter-scala](https://github.com/plokhotnyuk/jsoniter-scala) or/and start sponsoring my GitHub account to support its development and adoption of its tricks here.

**Comment by lbialy** [comment]
might have been an artifact of my jenv then! let me see

**Comment by lbialy** [comment]
```
[info] - java.time
[info]   - Decoder Sad Path
[info]     - DayOfWeek
[info]       (enum case not found: FOODY) did not satisfy equalTo((No enum constant java.time.DayOfWeek.FOODY))
[info]       (enum case not found: FOODY) did not satisfy (equalTo((No enum constant java.time.DayOfWeek.FOODY)) || equalTo((Unrecognized day of week name: FOODY)))
[info]       `stringify("foody").fromJson[DayOfWeek]` = Left((enum case not found: FOODY)) did not satisfy isLeft((equalTo((No enum constant java.time.DayOfWeek.FOODY)) || equalTo((Unrecognized day of week name: FOODY))))
[info]       at Some(/home/enigma/Projects/dev/zio-json/zio-json/shared/src/test/scala/zio/json/JavaTimeSpec.scala):346
[info]       (enum case not found: FOODY) did not satisfy equalTo((Unrecognized day of week name: FOODY))
[info]       (enum case not found: FOODY) did not satisfy (equalTo((No enum constant java.time.DayOfWeek.FOODY)) || equalTo((Unrecognized day of week name: FOODY)))
[info]       `stringify("foody").fromJson[DayOfWeek]` = Left((enum case not found: FOODY)) did not satisfy isLeft((equalTo((No enum constant java.time.DayOfWeek.FOODY)) || equalTo((Unrecognized day of week name: FOODY))))
[info]       at Some(/home/enigma/Projects/dev/zio-json/zio-json/shared/src/test/scala/zio/json/JavaTimeSpec.scala):346
[info]     - Month
[info]       (enum case not found: FEBTOBER) did not satisfy equalTo((No enum constant java.time.Month.FEBTOBER))
[info]       (enum case not found: FEBTOBER) did not satisfy (equalTo((No enum constant java.time.Month.FEBTOBER)) || equalTo((Unrecognized month name: FEBTOBER)))
[info]       `stringify("FebTober").fromJson[Month]` = Left((enum case not found: FEBTOBER)) did not satisfy isLeft((equalTo((No enum constant java.time.Month.FEBTOBER)) || equalTo((Unrecognized month name: FEBTOBER))))
[info]       at Some(/home/enigma/Projects/dev/zio-json/zio-json/shared/src/test/scala/zio/json/JavaTimeSpec.scala):1409
[info]       (enum case not found: FEBTOBER) did not satisfy equalTo((Unrecognized month name: FEBTOBER))
[info]       (enum case not found: FEBTOBER) did not satisfy (equalTo((No enum constant java.time.Month.FEBTOBER)) || equalTo((Unrecognized month name: FEBTOBER)))
[info]       `stringify("FebTober").fromJson[Month]` = Left((enum case not found: FEBTOBER)) did not satisfy isLeft((equalTo((No enum constant java.time.Month.FEBTOBER)) || equalTo((Unrecognized month name: FEBTOBER))))
[info]       at Some(/home/enigma/Projects/dev/zio-json/zio-json/shared/src/test/scala/zio/json/JavaTimeSpec.scala):1409
[info] Done
[error] Failed tests:
[error]         testzio.json.JavaTimeSpec
```

run for 3.1.2, sbt:
```
λ sbt -mem 8128 -v ++test
[sbt_options] declare -a sbt_options=([0]="-XX:MetaspaceSize=256M")
[process_args] java_version = '14'
[copyRt] java9_rt = '/home/enigma/.sbt/1.0/java9-rt-ext-adoptopenjdk_14_0_2/rt.jar'
# Executing command line:
java
-XX:+UseParallelGC
-Dfile.encoding=UTF8
-Xms8128m
-Xmx8128m
-Xss4M
-XX:ReservedCodeCacheSize=512m
-Dsbt.script=/usr/bin/sbt
-Dscala.ext.dirs=/home/enigma/.sbt/1.0/java9-rt-ext-adoptopenjdk_14_0_2
-XX:MetaspaceSize=256M
-jar
/usr/share/sbt/bin/sbt-launch.jar
++test

[info] welcome to sbt 1.6.2 (AdoptOpenJDK Java 14.0.2)
```

**Pull request merged**